File: bindings/jvm/src/main/java/io/github/wiltaylor/wcl/wasm/WasmCallbackBridge.java
package io.github.wiltaylor.wcl.wasm;

import io.github.wiltaylor.wcl.eval.WclValue;

import java.util.Map;
import java.util.function.Function;

public final class WasmCallbackBridge {
    private static final ThreadLocal<Map<String, Function<WclValue[], WclValue>>> FUNCTIONS =
            new ThreadLocal<>();

    private WasmCallbackBridge() {}

    public static void setFunctions(Map<String, Function<WclValue[], WclValue>> functions) {
        FUNCTIONS.set(functions);
    }

    public static void clearFunctions() {
        FUNCTIONS.remove();
    }

    public static CallResult invoke(String name, String argsJson) {
        var functions = FUNCTIONS.get();
        if (functions == null || !functions.containsKey(name)) {
            return new CallResult(false, "callback not found: " + name);
        }

        try {
            var argsNode = JsonConvert.MAPPER.readTree(argsJson);

            var args = new WclValue[argsNode.size()];
            for (int i = 0; i < argsNode.size(); i++) {
                args[i] = JsonConvert.toWclValue(argsNode.get(i));
            }

            var result = functions.get(name).apply(args);
            var resultJson = JsonConvert.wclValueToJson(result);
            return new CallResult(true, resultJson);
        } catch (Exception ex) {
            return new CallResult(false, ex.getMessage());
        }
    }

    public record CallResult(boolean success, String resultJson) {}
}

File: bindings/dotnet/src/Wcl/Wasm/WasmCallbackBridge.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using Wcl.Eval;

namespace Wcl.Wasm
{
    internal static class WasmCallbackBridge
    {
        [ThreadStatic]
        private static Dictionary<string, Func<WclValue[], WclValue>>? _functions;

        internal static void SetFunctions(Dictionary<string, Func<WclValue[], WclValue>> functions)
        {
            _functions = functions;
        }

        internal static void ClearFunctions()
        {
            _functions = null;
        }

        internal static (bool Success, string? ResultJson) Invoke(string name, string argsJson)
        {
            if (_functions == null || !_functions.TryGetValue(name, out var fn))
            {
                return (false, $"callback not found: {name}");
            }

            try
            {
                using var doc = JsonDocument.Parse(argsJson);
                var argsArray = doc.RootElement;

                var args = new WclValue[argsArray.GetArrayLength()];
                int i = 0;
                foreach (var el in argsArray.EnumerateArray())
                {
                    args[i++] = JsonConvert.ToWclValue(el);
                }

                var result = fn(args);
                var resultJson = JsonConvert.WclValueToJson(result);
                return (true, resultJson);
            }
            catch (Exception ex)
            {
                return (false, ex.Message);
            }
        }
    }
}
